be/src/formats/parquet/stored_column_reader.cpp

```diff
@@ -656,4 +674,256 @@ void StoredColumnReader::update_read_context(size_t records_read) {
     }
 }
 
+size_t StoredColumnReader::get_level_batch_size(size_t row, size_t num_values_left_in_cur_page, size_t decoded,
+                                                size_t parsed) {
+    constexpr size_t min_level_batch_size = 4096;
+    constexpr size_t max_level_batch_size = 1024 * 1024;
+    size_t levels_remaining = decoded - parsed;
+    if (row <= levels_remaining) {
+        return 0;
+    }
+
+    size_t levels_to_decode = std::max(min_level_batch_size, row - levels_remaining);
+    levels_to_decode = std::min(levels_to_decode, num_values_left_in_cur_page - levels_remaining);
+    levels_to_decode = std::min(levels_to_decode, max_level_batch_size);
+    return levels_to_decode;
+}
+
+size_t StoredColumnReader::count_not_null(level_t* def_levels, size_t num_parsed_levels, level_t max_def_level) {
+    size_t count = 0;
+    for (int i = 0; i < num_parsed_levels; ++i) {
+        level_t def_level = def_levels[i];
+        if (def_level == max_def_level) {
+            count++;
+        }
+    }
+    return count;
+}
+
+Status StoredColumnReader::read_range(const Range<uint64_t>& range, const Filter* filter,
```

Comment:
I suggest let's create more abstraction class, such as `PageIndexer`, so we can store and manipulate some member variables in these class instead of storing all of them in `StoredColumnReader`
```
PageIndexer _page_indexer;

read_range(start, end, filter) {
   _page_indexer.seek(start);
   while (row_read < row_count) {
       page = _page_indexer.next_page();
       read_page(page, filter);
   }
}
```